src/Connections.Abstractions/DefaultConnectionContext.cs

```diff
@@ -1,55 +1,99 @@
-            _features.Fetch(ref _features.Cache.ConnectionId, _ => null);
-
-        private IConnectionTransportFeature ConnectionTransportFeature =>
-            _features.Fetch(ref _features.Cache.ConnectionTransport, _ => null);
+        /// <summary>
+        /// Creates the DefaultConnectionContext without Pipes to avoid upfront allocations.
+        /// The caller is expected to set the <see cref="Transport"/> and <see cref="Application"/> pipes manually.
+        /// </summary>
+        /// <param name="id"></param>
+        public DefaultConnectionContext(string id)
+        {
+            ConnectionId = id;
 
-        private IConnectionItemsFeature ConnectionItemsFeature =>
-            _features.Fetch(ref _features.Cache.ConnectionItems, _ => null);
+            // PERF: This type could just implement IFeatureCollection
+            Features = new FeatureCollection();
+            Features.Set<IConnectionUserFeature>(this);
+            Features.Set<IConnectionItemsFeature>(this);
+            Features.Set<IConnectionIdFeature>(this);
+            Features.Set<IConnectionTransportFeature>(this);
+            Features.Set<IApplicationTransportFeature>(this);
+            Features.Set<IConnectionHeartbeatFeature>(this);
+        }
 
-        public override string ConnectionId
+        public DefaultConnectionContext(string id, IDuplexPipe transport, IDuplexPipe application)
+            : this(id)
         {
-            get => ConnectionIdFeature.ConnectionId;
-            set => ConnectionIdFeature.ConnectionId = value;
+            Transport = transport;
+            Application = application;
         }
 
-        public override IFeatureCollection Features => _features.Collection;
+        public override string ConnectionId { get; set; }
 
-        public override IDuplexPipe Transport
-        {
-            get => ConnectionTransportFeature.Transport;
-            set => ConnectionTransportFeature.Transport = value;
-        }
+        public override IFeatureCollection Features { get; }
+
+        public ClaimsPrincipal User { get; set; }
+
+        public override IDictionary<object, object> Items { get; set; } = new ConnectionItems();
 
-        public override IDictionary<object, object> Items
+        public IDuplexPipe Application { get; set; }
+
+        public override IDuplexPipe Transport { get; set; }
+
+        public void OnHeartbeat(Action<object> action, object state)
         {
-            get => ConnectionItemsFeature.Items;
-            set => ConnectionItemsFeature.Items = value;
+            lock (_heartbeatLock)
+            {
+                if (_heartbeatHandlers == null)
+                {
+                    _heartbeatHandlers = new List<(Action<object> handler, object state)>();
+                }
+                _heartbeatHandlers.Add((action, state));
+            }
         }
 
-        struct FeatureInterfaces
+        public void TickHeartbeat()
```

Comment:
It's the "host" (i.e. the actor that creates the `DefaultConnectionContext`) that ticks the heartbeat, so they need a mechanism to do this. They hold a reference to the concrete type, so they don't need the interface. Perhaps this could be specified better though.

Reply:
https://github.com/aspnet/SignalR/blob/077b17cad87c35fbce74b1b8cf4e53ce3955e79f/src/Microsoft.AspNetCore.Sockets.Http/ConnectionManager.cs#L180

Reply:
I’ll remove it. I have another SignalR PR that clears up these questions